src/main/java/uk/ac/sanger/sccp/stan/service/SampleProcessingServiceImp.java

```diff
@@ -0,0 +1,100 @@
+package uk.ac.sanger.sccp.stan.service;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Service;
+import uk.ac.sanger.sccp.stan.model.*;
+import uk.ac.sanger.sccp.stan.repo.LabwareRepo;
+import uk.ac.sanger.sccp.stan.repo.OperationTypeRepo;
+import uk.ac.sanger.sccp.stan.repo.SampleRepo;
+import uk.ac.sanger.sccp.stan.repo.TissueRepo;
+import uk.ac.sanger.sccp.stan.request.AddExternalIDRequest;
+import uk.ac.sanger.sccp.stan.request.OperationResult;
+
+import java.util.*;
+
+import static java.util.Objects.requireNonNull;
+
+@Service
+public class SampleProcessingServiceImp implements SampleProcessingService {
+    private final SampleRepo sampleRepo;
+    private final TissueRepo tissueRepo;
+    private final OperationTypeRepo opTypeRepo;
+    private final LabwareRepo labwareRepo;
+    private final OperationService opService;
+    private final Validator<String> externalNameValidator;
+
+    @Autowired
+    public SampleProcessingServiceImp(SampleRepo sampleRepo,
+                                      TissueRepo tissueRepo,
+                                      OperationTypeRepo opTypeRepo,
+                                      LabwareRepo labwareRepo,
+                                      OperationService opService,
+                                      @Qualifier("externalNameValidator") Validator<String> externalNameValidator) {
+        this.sampleRepo = sampleRepo;
+        this.tissueRepo = tissueRepo;
+        this.opTypeRepo = opTypeRepo;
+        this.labwareRepo = labwareRepo;
+        this.opService = opService;
+        this.externalNameValidator = externalNameValidator;
+    }
+
+    @Override
+    public OperationResult addExternalID(User user, AddExternalIDRequest request) {
+        requireNonNull(user, "User is null.");
+        requireNonNull(request, "Request is null.");
+        Collection<String> problems = new LinkedHashSet<>();
+
+        Labware lw = labwareRepo.getByBarcode(request.getLabwareBarcode());
+        String externalName = request.getExternalName();
+        List<Sample> samples = new ArrayList<>();
```

Comment:
I suggest you use a set for this. It should not be a problem if the labware contains one sample multiple times.
```
Set<Sample> samples = lw.getSlots().stream()
    .flatMap(slot -> slot.getSamples().stream())
    .collect(toSet());
```